File: code/hashing/custom-hash.cpp
// A custom hash for your own key type combines the fields
// so that (1,2) and (2,1) differ. Equal keys must hash equally.
struct PointHash {
    size_t operator()(const Point& p) const {
        return hash<int>()(p.x) * 31 + hash<int>()(p.y);
    }
};
unordered_map<Point, int, PointHash> m;

File: code/hashing/custom-hash.rs
// derive(Hash) combines the fields automatically; Eq is required
// alongside it. Equal keys hash equally by construction.
use std::collections::HashMap;

#[derive(PartialEq, Eq, Hash)]
struct Point { x: i32, y: i32 }

let mut m: HashMap<Point, i32> = HashMap::new();
